uk_geo_utils/base_updater.py

```diff
@@ -0,0 +1,377 @@
+
+        for statement in index_rename_statements:
+            self.stdout.write(f"Executing: {statement}")
+            self.cursor.execute(statement)
+
+    def rename_temp_table(self):
+        self.stdout.write("Renaming temp table...")
+        rename_table_statement = (
+            f"ALTER TABLE {self.temp_table_name} RENAME TO {self.table_name}"
+        )
+        self.stdout.write(f"Executing: {rename_table_statement}")
+        self.cursor.execute(rename_table_statement)
+        self.stdout.write("Renaming primary key...")
+        primary_key_rename_statement = f"ALTER TABLE {self.table_name} RENAME CONSTRAINT {self.primary_key_constraint['temp_name']} TO {self.primary_key_constraint['constraint_name']}"
+        self.stdout.write(f"Executing: {primary_key_rename_statement}")
+        self.cursor.execute(primary_key_rename_statement)
+
+        index_rename_statements = []
+        for index in self.indexes:
+            if self.index_exists(index["temp_index_name"]):
+                index_rename_statements.append(
+                    index["rename_temp_index_statement"]
+                )
+
+        for statement in index_rename_statements:
+            self.stdout.write(f"Executing: {statement}")
+            self.cursor.execute(statement)
+
+    def get_constraints_and_index_statements(self):
+        self.stdout.write(
+            f"Getting constraints and indexes for {self.table_name}"
+        )
+        self.primary_key_constraint = self.get_primary_key_constraint()
+        self.indexes = self.get_index_statements()
+        self.foreign_key_constraints = self.get_foreign_key_constraints()
+
+    def handle(self, **options):
+        if not check_memory():
+            raise Exception(
+                "This instance has less than the recommended memory. Try running the import from a larger instance."
+            )
+
+        self.get_data_path(options)
+
+        self.get_constraints_and_index_statements()
+
+        try:
+            # Create empty temp tables
+            self.create_temp_table()
+
+            # import data into the temp table
+            self.import_data_to_temp_table()
+
+            # Add temp primary keys
+            self.add_temp_primary_key()
+
+            # Add temp indexes
+            self.build_temp_indexes()
+
+            with transaction.atomic():
+                # Drop Foreign keys
+                if self.foreign_key_constraints:
+                    self.drop_foreign_keys()
+
+                # Rename old tables, pkeys and indexes (add _old)
+                self.rename_old_table()
+
+                # Rename temp table to original names, pkey and indexes
+                self.rename_temp_table()
+
+                # Add Foreign keys
+                if self.foreign_key_constraints:
+                    self.add_foreign_keys()
+
+            # Validate foreign key constraints
+            if self.foreign_key_constraints:
+                self.validate_foreign_keys()
```

Comment:
Should `validate_foreign_keys()` happen inside the transaction?
i.e: if it fails, do we want to roll back the whole thing?
..or do we have to commit the transaction before we can do this?

Reply:
The reason they don't is that the transaction would then be blocking while the foreign keys are validated. In fairness it doesn't take that long, so maybe it's a reasonable trade off. In that case the create statements can just loose the `NOT VALID` clause on the end.

Reply:
Notes from call: Lets try moving `validate_foreign_keys()` inside the transaction and bin the `_old` tables/indexes.

Reply:
Foreign key validation inside transaction: a49226f1fe59c
old tables removed: 978bf3fc23b35b (referenced in previous comment)